### PR TITLE
BLD: Meson `__config__` generation

### PR DESCRIPTION
**Pull request by ganesh-k13**
### Changes
Add functionality to autogenerate build information from Meson information
In order to add new information, do the following:
* Add the information as an argument in `numpy/meson.build`
* Modify `__config__.py.in` to accept the new argument

### New `__config__.py`:
<details>
<summary> 'print' mode </summary>
<br>

```py
In [2]: np.show_config()                                                                                                                                                                          
Build Dependencies:                                                                                                                                                                               
  blas:                                                                                                                                                                                           
    found: 'True'                                                                                                                                                                                 
    includedir: /usr/include/x86_64-linux-gnu/openblas-pthread/                                                                                                                                   
    libdir: /usr/lib/x86_64-linux-gnu/openblas-pthread/                                                                                                                                           
    name: openblas                                                                                                                                                                                
    openblas config: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS= NO_LAPACK=                                                                                                            
      NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64                                                                                                                              
    pc file directory: /usr/lib/x86_64-linux-gnu/pkgconfig
    type name: pkgconfig                                                                         
    version: 0.3.20                                                                              
  lapack:                                                                                                                                                                                         
    found: 'True'                                                                                
    includedir: /usr/include/x86_64-linux-gnu/openblas-pthread/                                  
    libdir: /usr/lib/x86_64-linux-gnu/openblas-pthread/
    name: openblas                                                                               
    openblas_config: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS= NO_LAPACK=           
      NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64                             
    pc file directory: /usr/lib/x86_64-linux-gnu/pkgconfig
    type_name: pkgconfig                                                                         
    version: 0.3.20                                                                              
Compilers:                                                                                                                                                                                        
  c:                                                                                             
    commands: cc                                
    linker: ld.bfd                              
    name: gcc                                   
    version: 11.3.0                                                                              
  c++:                                          
    commands: c++                               
    linker: ld.bfd                              
    name: gcc                                   
    version: 11.3.0                                                                              
  cython:                                       
    commands: cython                                                                             
    linker: cython                                                                               
    name: cython                                                                                 
    version: 0.29.30                                                                             
Machine Information:                                                                             
  build:                                                                                         
    cpu: x86_64 
    endian: little
    family: x86_64
    system: linux
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /home/ganesh/os/np-test/bin/python3
  version: '3.10'
```
</details>

<details>
<summary> 'json' mode </summary>
<br>

```py
In [3]: np.show_config(mode='json')
Out[3]: 
{'Compilers': {'c': {'name': 'gcc',
   'linker': 'ld.bfd',
   'version': '11.3.0',
   'commands': 'cc'},
  'cython': {'name': 'cython',
   'linker': 'cython',
   'version': '0.29.30',
   'commands': 'cython'},
  'c++': {'name': 'gcc',
   'linker': 'ld.bfd',
   'version': '11.3.0',
   'commands': 'c++'}},
 'Machine Information': {'host': {'cpu': 'x86_64',
   'family': 'x86_64',
   'endian': 'little',
   'system': 'linux'},
  'build': {'cpu': 'x86_64',
   'family': 'x86_64',
   'endian': 'little',
   'system': 'linux'}},
 'Build Dependencies': {'blas': {'name': 'openblas',
   'found': 'True',
   'version': '0.3.20',
   'type name': 'pkgconfig',
   'includedir': '/usr/include/x86_64-linux-gnu/openblas-pthread/',
   'libdir': '/usr/lib/x86_64-linux-gnu/openblas-pthread/',
   'openblas config': 'USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS= NO_LAPACK= NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64',
   'pc file directory': '/usr/lib/x86_64-linux-gnu/pkgconfig'},
  'lapack': {'name': 'openblas',
   'found': 'True',
   'version': '0.3.20',
   'type_name': 'pkgconfig',
   'includedir': '/usr/include/x86_64-linux-gnu/openblas-pthread/',
   'libdir': '/usr/lib/x86_64-linux-gnu/openblas-pthread/',
   'openblas_config': 'USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS= NO_LAPACK= NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64',
   'pc file directory': '/usr/lib/x86_64-linux-gnu/pkgconfig'}},
 'Python Information': {'path': '/home/ganesh/os/np-test/bin/python3',
  'version': '3.10'}}

```
</details>

<details>
<summary> 'print' mode no PyYaml </summary>
<br>

```py
In [2]: np.show_config()                                                                         
/home/ganesh/os/numpy/build-install/usr/lib/python3/dist-packages/numpy/__config__.py:117: UserWarning: Install `pyyaml` for better output                  
  warnings.warn('Install `pyyaml` for better output')           
{                                               
  "Compilers": {                                
    "c": {                                      
      "name": "gcc",                            
      "linker": "ld.bfd", 
      "version": "11.3.0",     
      "commands": "cc"                                                                           
    },                                                                                           
    "cython": {                                                                                                                                                                                   
      "name": "cython",                                                                          
      "linker": "cython",                       
      "version": "0.29.30",                     
      "commands": "cython"
    },                                                                                           
    "c++": {                                    
      "name": "gcc",                            
      "linker": "ld.bfd",                       
      "version": "11.3.0",
      "commands": "c++"                         
    }                                           
  },                                            
  "Machine Information": {
    "host": {                                   
      "cpu": "x86_64",                          
      "family": "x86_64",
      "endian": "little",
      "system": "linux"                         
    },
    "build": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "linux"
    }
  },
  "Build Dependencies": {
    "blas": {
      "name": "openblas",
      "found": "True",
      "version": "0.3.20",
      "type name": "pkgconfig",
      "includedir": "/usr/include/x86_64-linux-gnu/openblas-pthread/",
      "libdir": "/usr/lib/x86_64-linux-gnu/openblas-pthread/",
      "openblas config": "USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS= NO_LAPACK= NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64",
      "pc file directory": "/usr/lib/x86_64-linux-gnu/pkgconfig"
    },
    "lapack": {
      "name": "openblas",
      "found": "True",
      "version": "0.3.20",
      "type_name": "pkgconfig",
      "includedir": "/usr/include/x86_64-linux-gnu/openblas-pthread/",
      "libdir": "/usr/lib/x86_64-linux-gnu/openblas-pthread/",
      "openblas_config": "USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS= NO_LAPACK= NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64",
      "pc file directory": "/usr/lib/x86_64-linux-gnu/pkgconfig"
    }
  },
  "Python Information": {
    "path": "/home/ganesh/os/np-test/bin/python3",
    "version": "3.10"
  }
}
```
</details>

### TODO

- [ ] ~Remove `__config__.py.in`~
- [x] Add `Build Dependencies`
  - [x] BLAS
  - [x] LAPACK
- [x] Add tests (surprised nothing failed after this change)
- [x] Add SIMD
- [x] Add release notes
- [x] Add docs

related: https://github.com/numpy/numpy/pull/20939#issuecomment-1336666340 | https://github.com/scipy/scipy/issues/16803

### CODE REVIEW

**Comment by ganesh-k13** [comment]
Hi @rgommers , I've made a POC of what I had in mind and wanted to know if this is the right approach for the requirement. I can continue based on feedback.

**Comment by eli-schwartz** [comment]
Since this only contains information known at the time of configuration, I wonder if it's necessary to have a generator script at all? Maybe passing a dict in meson to a `configure_file()` and using that to fill out a template, would work equally well?

The main difference is that it requires less work and one fewer build edge.

**Comment by ganesh-k13** [comment]
Ah ok that makes sense. Something like [this](https://github.com/scipy/scipy/blob/57ed5c01e01a6d261c68744d3c16a0364882e5a3/scipy/_lib/meson.build#L77-L88) yeah?

**Comment by rgommers** [comment]
> Ah ok that makes sense. Something like [this](https://github.com/scipy/scipy/blob/57ed5c01e01a6d261c68744d3c16a0364882e5a3/scipy/_lib/meson.build#L77-L88) yeah?

That example is not the best (the if-else could use a single `.set10` call instead), but the basic principle is okay for a dict passed to `configure_file`. In NumPy itself there's a more comprehensive example, including use of a template: https://github.com/numpy/numpy/blob/a24ddad5d1d43ec9bea474abf8068ee6b8db01db/numpy/core/meson.build#L423

File: numpy/meson.build [hunk 1]
```diff
@@ -143,12 +145,45 @@ foreach subdir: pure_subdirs
...
+  conf_data.set(name + '_COMP_CMD_ARRAY', compiler.cmd_array())
+endforeach
+
+# Host CPU and system information
+conf_data.set('HOST_CPU', host_machine.cpu())
+conf_data.set('HOST_CPU_FAMILY', host_machine.cpu_family())
+conf_data.set('HOST_CPU_ENDIAN', host_machine.endian())
+conf_data.set('HOST_CPU_SYSTEM', host_machine.system())
+
+# Python information
```
Comment (ganesh-k13) [inline numpy/meson.build:168]: We could perhaps add some more info here using https://mesonbuild.com/Python-module.html

File: numpy/meson.build [hunk 2]
```diff
@@ -143,12 +145,45 @@ foreach subdir: pure_subdirs
   install_subdir(subdir, install_dir: np_dir)
 endforeach
 
-custom_target('__config__.py',
-  output: '__config__.py',
+cc = meson.get_compiler('c')
+cpp = meson.get_compiler('cpp')
```
Comment (rgommers) [inline numpy/meson.build:149]: These two lines aren't needed, the `cc` and `cpp` variables are already defined in the top-level `meson.build`

File: numpy/meson.build [hunk 3]
```diff
@@ -143,12 +145,45 @@ foreach subdir: pure_subdirs
...
+compilers = {'C': cc, 'CPP': cpp}
```
Comment (rgommers) [inline numpy/meson.build:150]: `cython` too perhaps?

File: numpy/meson.build [hunk 4]
```diff
@@ -143,12 +145,45 @@ foreach subdir: pure_subdirs
...
+conf_data.set('HOST_CPU_SYSTEM', host_machine.system())
+
+# Python information
+conf_data.set('PYTHON_PATH', py.path())
+
+# Dependencies information
+foreach name, dep : dependency_map
+  conf_data.set(name + '_NAME', dep.name())
+  conf_data.set(name + '_FOUND', dep.found())
+  conf_data.set(name + '_VERSION', dep.version())
```
Comment (rgommers) [inline numpy/meson.build:175]: Probably all the info from this line on should be gated on `dep.found()` being `true`?

`dep.type_name()` may be useful to add.

**Comment by rgommers** [comment]
This looks pretty good at first glance @ganesh-k13. I added some comments, will do a proper review later.

File: numpy/meson.build (hunk 1 above)
Comment (ganesh-k13) [reply numpy/meson.build:170]: Added on [a4006f2](https://github.com/numpy/numpy/pull/22769/commits/a4006f2af87295a44808fb1a927f7561909688d9)

File: numpy/meson.build (hunk 4 above)
Comment (ganesh-k13) [reply numpy/meson.build]: Added type_name on [a4006f2](https://github.com/numpy/numpy/pull/22769/commits/a4006f2af87295a44808fb1a927f7561909688d9) and `if` check on [525f0bc](https://github.com/numpy/numpy/pull/22769/commits/525f0bc6e7cd14960bfc3853fb83b3dc04bd0644)

File: numpy/meson.build (hunk 2 above)
Comment (ganesh-k13) [reply numpy/meson.build]: refined on [766dcf4](https://github.com/numpy/numpy/pull/22769/commits/766dcf44ed586b6e738eaf86f963ab2740ea4c37)

**Comment by ganesh-k13** [comment]
Are we planning on using [Meson Tests](https://mesonbuild.com/Unit-tests.html)? I was exploring how to add UT. If it's in the plan, I can start the test framework with this module, else I'll stick with pytest and see how to test this.

**Comment by eli-schwartz** [comment]
If you've already heavily invested in pytest, meson test doesn't add much. Particularly, there's not a whole lot of benefit in taking one test harness/reporter (meson test) and using it to run another one (pytest).

You can do it, but it will just result in meson test reporting on a single test (that is, whether running `pytest` passes or fails) and then you can inspect the meson test logs to see the output of pytest.

**Review by rgommers — CHANGES_REQUESTED** [review]
Hi @ganesh-k13, I did some testing. It all seems to work, my main observation is that the output is quite difficult to read. For comparison, let's have a look at `show_runtime`, `show_config` and the output of Meson itself:
```
In [2]: np.show_runtime()
[{'simd_extensions': {'baseline': [], 'found': [], 'not_found': []}},
 {'architecture': 'SkylakeX',
  'filepath': '/home/rgommers/mambaforge/envs/numpy-dev/lib/libopenblasp-r0.3.21.so',
  'internal_api': 'openblas',
  'num_threads': 24,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.21'}]

In [3]: np.show_config()
Out[3]: 
{'C Compiler': {'C_COMP': 'gcc',
  'C_COMP_LINKER_ID': 'ld.bfd',
  'C_COMP_VERSION': '10.4.0',
...
```
And the Meson configure output:
```
C compiler for the host machine: /home/rgommers/mambaforge/envs/numpy-dev/bin/x86_64-conda-linux-gnu-cc (gcc 10.4.0 "x86_64-conda-linux-gnu-cc (conda-forge gcc 10.4.0-19) 10.4.0")
C linker for the host machine: /home/rgommers/mambaforge/envs/numpy-dev/bin/x86_64-conda-linux-gnu-cc ld.bfd 2.39
C++ compiler for the host machine: /home/rgommers/mambaforge/envs/numpy-dev/bin/x86_64-conda-linux-gnu-c++ (gcc 10.4.0 "x86_64-conda-linux-gnu-c++ (conda-forge gcc 10.4.0-19) 10.4.0")
C++ linker for the host machine: /home/rgommers/mambaforge/envs/numpy-dev/bin/x86_64-conda-linux-gnu-c++ ld.bfd 2.39
Cython compiler for the host machine: cython (cython 0.29.32)
Host machine cpu family: x86_64
Host machine cpu: x86_64
...
```

The capitals, hard to understand names and returning a dict of dicts instead of actually printing all contribute to the output being tricky for the reader to parse. I suggest changing the keys of the dict being as easy as possible to understand. And then also providing an option to print to stdout rather than returning a dict of dicts.

A few examples to make that concrete:
```
'BLAS_VERSION': '0.3.21',
```
I'd make that `'version': '0.3.21'`  (it's already under a BLAS key).

```
'LAPACK_TYPE_NAME': 'pkgconfig',
```
I'd make this: `'detection method': 'pkgconfig',`

**Comment by ganesh-k13** [comment]
Yeah, that's a really good point! In fact Matti had a similar idea in the previous PR(https://github.com/numpy/numpy/pull/20939#issuecomment-1025116089). I'll make those changes 👍

**Comment by ganesh-k13** [comment]
I have added `modes`, please see the issue description for examples.

File: numpy/meson.build [hunk 5]
```diff
@@ -143,12 +145,62 @@ foreach subdir: pure_subdirs
...
+  'CPP': cpp,
+  'CYTHON': meson.get_compiler('cython')
+}
+
+machines = {
+  'HOST': host_machine,
+  'BUILD': build_machine,
+}
+
+if meson.is_cross_build()
```
Comment (ganesh-k13) [inline numpy/meson.build:159]: Not sure how useful this is, please let me know if we need it.

File: numpy/__config__.py.in [hunk 6]
```diff
@@ -1,134 +1,125 @@
...
+    if mode == DisplayModes.print.value:
+        try:  # Non-standard library, check import
+            import yaml
+
+            print(yaml.dump(CONFIG))
+        except ModuleNotFoundError:
+            import warnings
+            import json
+
+            warnings.warn("Install `pyyaml` for better output")
```
Comment (ganesh-k13) [inline numpy/__config__.py.in:120]: ah this needs a `level`, will add.

File: numpy/meson.build (hunk 5 above)
Comment (rgommers) [reply numpy/meson.build:159]: That does seem useful to me.

File: numpy/__config__.py.in [hunk 7]
```diff
@@ -1,134 +1,125 @@
...
+                "endian": "@TARGET_CPU_ENDIAN@",
+                "system": "@TARGET_CPU_SYSTEM@",
+            },
+        },
+        "Build Dependencies": {
+            "blas": {
+                "name": "@BLAS_NAME@",
+                "found": "@BLAS_FOUND@",
+                "version": "@BLAS_VERSION@",
+                "type name": "@BLAS_TYPE_NAME@",
```
Comment (rgommers) [inline numpy/__config__.py.in:76]: I propose "detection method". It's independent of BLAS/LAPACK itself, so "type name" seems a little confusing.

File: numpy/__config__.py.in [hunk 8]
```diff
@@ -1,134 +1,125 @@
...
+                "includedir": "@BLAS_INCLUDEDIR@",
+                "libdir": "@BLAS_LIBDIR@",
+                "openblas config": "@BLAS_OPENBLAS_CONFIG@",
+                "pc file directory": "@BLAS_PCFILEDIR@",
+            },
+            "lapack": {
+                "name": "@LAPACK_NAME@",
+                "found": "@LAPACK_FOUND@",
+                "version": "@LAPACK_VERSION@",
+                "type_name": "@LAPACK_TYPE_NAME@",
```
Comment (rgommers) [inline numpy/__config__.py.in:86]: Same here, "detection method"

File: numpy/__config__.py.in [hunk 9]
```diff
@@ -1,134 +1,125 @@
...
+            ],
+            "not found": [
+                feature for feature in __cpu_dispatch__ if not __cpu_features__[feature]
+            ],
+        },
+    }
+)
+
+
+def show(mode="print"):
```
Comment (rgommers) [inline numpy/__config__.py.in:110]: This function still needs some docs.

**Comment by ganesh-k13** [comment]
Overall I think we can refine the keys a bit to make more sense. I'm torn between making it programmatically accessible vs human readable, for example: `openblas_config` vs `openblas configuration`. Please feel free to suggest any changes to keys or the way we are displaying them if PyYaml dump makes sense or we can write our own `tabulate` of sorts.

File: numpy/__config__.py.in (hunk 9 above)
Comment (rgommers) [reply numpy/__config__.py.in:110]: Regarding the mode names, `print` is a builtin` and `json` isn't really what is returned. How about:
- `print` -> `stdout` 
- `json` -> `dicts`

**Review by rgommers — COMMENTED** [review]
This is getting close! Just a few minor comments, and docs left to be added.

**Comment by rgommers** [comment]
> Overall I think we can refine the keys a bit to make more sense. I'm torn between making it programmatically accessible vs human readable, for example: `openblas_config` vs `openblas configuration`.

I think optimizing for human-readable when printing sounds fine to me. I think that's the first thing people will do. Programmatic access is important too, but it's fine is that requires say a space in the key of a dict.

File: numpy/__config__.py.in (hunk 7 above)
Comment (ganesh-k13) [reply numpy/__config__.py.in:76]: Ah I missed this, thanks will add it

**Comment by ganesh-k13** [comment]
> I think optimizing for human-readable when printing sounds fine to me.
Thanks for the clarification, I'll work on that.

File: numpy/__config__.py.in (hunk 7 above)
Comment (ganesh-k13) [reply numpy/__config__.py.in]: Refined in 72a73e6dc5a64230f69cd61707f6ddd18eaf935c

File: numpy/__config__.py.in (hunk 8 above)
Comment (ganesh-k13) [reply numpy/__config__.py.in]: Refined in 72a73e6dc5a64230f69cd61707f6ddd18eaf935c

File: numpy/__config__.py.in (hunk 9 above)
Comment (ganesh-k13) [reply numpy/__config__.py.in]: Refined in 72a73e6dc5a64230f69cd61707f6ddd18eaf935c

File: numpy/tests/test_numpy_config.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
+"""
```
Comment (ganesh-k13) [inline numpy/tests/test_numpy_config.py:1]: I realised after pushing, this test should be restricted to meson builds for now. Will look into it.

File: numpy/__config__.py.in [hunk 11]
```diff
@@ -55,80 +143,23 @@ def show():
...
-    Examples
-    --------
-    >>> import numpy as np
-    >>> np.show_config()
-    blas_opt_info:
-        language = c
-        define_macros = [('HAVE_CBLAS', None)]
-        libraries = ['openblas', 'openblas']
-        library_dirs = ['/usr/local/lib']
+    1. 'stdout' mode requires ``PyYaml`` module for better output
```
Comment (rgommers) [inline numpy/__config__.py.in:146]: It's not required, right? I'd rephrase as:
```
The `'stdout'` mode will give more readable output if ``pyyaml`` is installed
```

**Comment by rgommers** [comment]
When you think this is ready to merge, please take it out of Draft mode and let me know - happy to get this in soon:)

**Comment by ganesh-k13** [comment]
I didn't realise it was draft. Yeah almost ready, I'll just restrict the tests. Does it need a release note as show_config takes args now? Or we just wait for distutils to go and add one big release note?

**Comment by rgommers** [comment]
distutils to go is probably ~5 months away, so let's not wait for that.

File: numpy/__config__.py.in (hunk 11 above)
Comment (ganesh-k13) [reply numpy/__config__.py.in]: Edited old commit 6935fe986dc2eb58d822d45463766fae3cfbc5e7

File: numpy/tests/test_numpy_config.py (hunk 10 above)
Comment (ganesh-k13) [reply numpy/tests/test_numpy_config.py:1]: Edited old commit to 36d865d97e201b54b14071dabcd8553ce72d5081

File: numpy/meson.build (hunk 5 above)
Comment (rgommers) [reply numpy/meson.build:159]: Actually, I have to retract that. For cross-compilation, `build` and `host` should cover things. We don't really have a need for `target`. It's current also generating this warning:
```
numpy/meson.build:199: WARNING: The variable(s) 'TARGET_CPU', 'TARGET_CPU_ENDIAN', 'TARGET_CPU_FAMILY', 'TARGET_CPU_SYSTEM' in the input file 'numpy/__config__.py.in' are not present in the given configuration data.
```

The current output is of `show_config` is:
```
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
```
That looks good. I'd suggest removing `target`, and adding one line:
```
cross-compiled: yes/no
```

Comment (rgommers) [reply numpy/meson.build:159]: There's one more warning, that should be easy to fix:
```
numpy/meson.build:182: WARNING: Project targets '>= 0.64.0' but uses feature deprecated since '0.55.0': ExternalProgram.path. use ExternalProgram.full_path() instead
Configuring __config__.py using configuration
```

Comment (ganesh-k13) [reply numpy/meson.build]: Fixed in a35447fac005c8794ef620949a0448a8f4715251

**Comment by ganesh-k13** [comment]
Hey @rgommers , any more changes needed on this PR?

**Review by rgommers — APPROVED** [review]
Thanks for the ping @ganesh-k13, I missed the last update. This looks good to go now, time to merge it. Thanks a lot for getting this over the finish line!

**Comment by seberg** [comment]
I am running into a circular import trying to add a flag to the main namespace.  I could change that design to be a getter or move the initialization around.

But I am wondering if we can move the C imports here back to be done lazily to avoid importing C until later?

**Comment by seberg** [comment]
Nvm, it is nice to `__cpu_baseline__` defined in that scope...  Maybe this should be cleaned up, but I think the solution here is to just move my thing before this init.

**Comment by ganesh-k13** [comment]
We could refine the cleanup to take place in the actual function and move the import inside. I think that's cleaner as well. Shall I do that? I feel we will hit this in the future.

**Comment by seberg** [comment]
I think it is fine as is, thanks.  I had added the `_utils` to move most things that may be required in C into.  Just now, I thought I would have this one flag directly in the main namespace, so it needs to happen before the import.

The only thing I could think of doing would be to make it a bit more obvious where the C-module is imported the first time (which is also the place guarded by this?):
```
    except ImportError as e:
        msg = """Error importing numpy: you should not try to import numpy from
        its source directory; please exit the numpy source tree, and relaunch
        your python interpreter from there."""
        raise ImportError(msg) from e
```
I think.  Maybe with a comment that things that used from C must be defined before or live live in `_utils/`.
